patches/api/0447-Provide-registered-listeners-in-the-PluginManager.patch

```diff
@@ -0,0 +1,244 @@
+ import org.jetbrains.annotations.NotNull;
+ import org.jetbrains.annotations.Nullable;
+@@ -115,25 +116,36 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+     /**
+      * Registers all the events in the given listener class
+      *
++     * Superseded by {@link PluginManager#registerListeners(Listener, Plugin)} which
++     * also returns the registered listener
++     *
+      * @param listener Listener to register
+      * @param plugin Plugin to register
+      */
++    @ApiStatus.Obsolete
+     public void registerEvents(@NotNull Listener listener, @NotNull Plugin plugin);
+ 
+     /**
+      * Registers the specified executor to the given event class
+      *
++     * Superseded by {@link PluginManager#registerListener(Class, Listener, EventPriority, EventExecutor, Plugin)}
++     * which also returns the registered listener
++     *
+      * @param event Event type to register
+      * @param listener Listener to register
+      * @param priority Priority to register this event at
+      * @param executor EventExecutor to register
+      * @param plugin Plugin to register
+      */
++    @ApiStatus.Obsolete
+     public void registerEvent(@NotNull Class<? extends Event> event, @NotNull Listener listener, @NotNull EventPriority priority, @NotNull EventExecutor executor, @NotNull Plugin plugin);
+ 
+     /**
+      * Registers the specified executor to the given event class
+      *
++     * Superseded by {@link PluginManager#registerListener(Class, Listener, EventPriority, EventExecutor, Plugin, boolean)}
++     * which also returns the registered listeners
++     *
+      * @param event Event type to register
+      * @param listener Listener to register
+      * @param priority Priority to register this event at
+@@ -141,6 +153,7 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+      * @param plugin Plugin to register
+      * @param ignoreCancelled Whether to pass cancelled events or not
+      */
++    @ApiStatus.Obsolete
+     public void registerEvent(@NotNull Class<? extends Event> event, @NotNull Listener listener, @NotNull EventPriority priority, @NotNull EventExecutor executor, @NotNull Plugin plugin, boolean ignoreCancelled);
+ 
+     /**
+@@ -316,5 +329,42 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+      */
+     @org.jetbrains.annotations.ApiStatus.Experimental
+     void overridePermissionManager(@NotNull Plugin plugin, @Nullable io.papermc.paper.plugin.PermissionManager permissionManager);
++
++    /**
++     * Registers all the events in the given listener class
++     *
++     * @param listener Listener to register
++     * @param plugin Plugin to register
++     * @return The new registered listeners, may be empty if no event handlers were found
++     */
++    @NotNull
++    public java.util.Collection<RegisteredListener> registerListeners(@NotNull Listener listener, @NotNull Plugin plugin);
```

Comment:
Naming this `registerListener` too might also be good? Provided that it only accepts one Listener and not multiple, I think naming this `registerListener` might be a good idea.
Alternatively, could this maybe accept a vararg Listener array?

Reply:
That's fair, I had it as `registerListeners` as it replaces `registerEvents`, but the new one still only registers one listener, even if it registers multiple events.